### PR TITLE
Fixes and Improvements to WCSAxes

### LINKED ISSUE #9370: WCSAxes no longer has an entry for every world dimension in CoordinatesMap
This was introduced in #8409 and to some extent is a feature as opposed to a bug, however, now if you pass a WCS to `WCSAxes` which has some separable dimensions, and you slice them out with the `slices` argument. (i.e. you plot the two spatial axes of a space, space, wavelength cube) there will not be an entry in the `WCSAxes.coords` object for the sliced out axes. (i.e. wavelength).

99% of the time this is fine, but is firstly a change in behaviour, and secondly, a pain when you are trying to cook up complex multi-dimensional plotting code like [this](https://github.com/sunpy/ndcube/pull/207/files#diff-8a5b90e322dfe13d80faa13b3836109aR202-R214).

I am opening this mainly because I want this choice to be very explicit before we ship it in 4.0.

@astrofrog

### PR DESCRIPTION
**Pull request by Cadair**
This PR is a collection of WCSAxes related fixes, primarily focusing on APE 14 and 1D plotting.

It does the following things:

- Fixes a bug in `WCSAxes` if `slices` is passed and `SlicedLowLevelWCS` is the wcs object. (https://github.com/Cadair/astropy/commit/f7cc7b8f2df4d7ec8a65caec48ca1f3074fca617#commitcomment-35522807)
- Adds a `get_format_unit` method to `CoordianateHelper`
- Raises an error in `SlicedLowLevelWCS` if the slice is too long.
- Fixes #9370 
- Puts the most rapidly varying world axis on the bottom spine when plotting a 1D plot.
- Automatically add labels with the coordinate type and unit to plots

### CODE REVIEW

File: astropy/wcs/utils.py [hunk 1]
```diff
@@ -632,3 +632,39 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
     coords = cls(frame.realize_frame(data))
 
     return coords
+
+
+def local_partial_pixel_derivatives(wcs, *pixel, normalize_by_world=False):
+    """
+    Return a matrix of shape ``(pixel_n_dim, world_n_dim)`` where each entry
```
Comment (Cadair) [inline astropy/wcs/utils.py]: @astrofrog This is transposed from APE 14, does it make sense to flip it?

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py [hunk 2]
```diff
@@ -217,6 +219,86 @@ def test_coord_type_1d_2d_wcs_uncorrelated():
...
+    coord_meta['visible'] = [False, True, True]
+    coord_meta['format_unit'] = [u.Unit("min"), u.Unit("arcsec"), u.Unit("arcsec")]
+    coord_meta['default_axislabel_position'] = ['', 'b', 't']
+    coord_meta['default_ticklabel_position'] = ['', 'b', 't']
+    coord_meta['default_ticks_position'] = ['', 'btlr', 'btlr']
+
+    with warnings.catch_warnings():
+        warnings.filterwarnings('ignore', category=FutureWarning)
+        # Validate the axes initialize correctly
+        ax = plt.subplot(projection=sub_wcs, slices=slices_wcsaxes)
```
Comment (astrofrog) [inline astropy/visualization/wcsaxes/tests/test_wcsapi.py]: So is there anything to check here or are you just making sure the initialization doesn't crash?

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py [hunk 3]
```diff
@@ -217,6 +219,86 @@ def test_coord_type_1d_2d_wcs_uncorrelated():
...
+                             (np.s_[:,:,0,:], [0, 'x', 'y'], (1, 2), 2),
+                             (np.s_[:,:,0,:], ['x', 0, 'y'], (0, 2), 3),
+                             (np.s_[:,:,0,:], ['x', 'y', 0], (0, 1), 3),
+                             (np.s_[:,0,:,:], ['x', 'y', 0], (0, 1), 2),
+                         ])
+def test_apply_slices(sub_wcs, wcs_slice, wcsaxes_slices, world_map, ndim):
+    transform_wcs, _, world_map = apply_slices(sub_wcs, wcsaxes_slices)
+    assert transform_wcs.world_n_dim == ndim
+
+    assert world_map == world_map
```
Comment (astrofrog) [inline astropy/visualization/wcsaxes/tests/test_wcsapi.py]: I think you might need to use ``_expected`` or something here!

File: astropy/visualization/wcsaxes/coordinate_helpers.py [hunk 4]
```diff
@@ -293,6 +293,12 @@ def set_format_unit(self, unit, decimal=None, show_decimal_unit=True):
         self._formatter_locator.decimal = decimal
         self._formatter_locator.show_decimal_unit = show_decimal_unit
 
+    def get_format_unit(self):
```
Comment (astrofrog) [inline astropy/visualization/wcsaxes/coordinate_helpers.py]: Can you add a test? (well just add a line to a test that does ``set_format_unit``

File: astropy/visualization/wcsaxes/tests/test_wcsapi.py (hunk 2 above)
Comment (Cadair) [reply astropy/visualization/wcsaxes/tests/test_wcsapi.py]: Just making sure the init doesn't crash.

**Comment by Cadair** [comment]
Given the addition of the auto spine selection for 1D plots, I have made a change here which by default plots the physical type (and for scalar axes unit) as an axis label. This primarily so that it is easy to identify which spatial axis is which, but is also obviously more generally useful.

This is a reasonably large change to the default behaviour, so I would be happy to add it only to the 1D plots if people think that's a good idea. @astrofrog @larrybradley

File: astropy/visualization/wcsaxes/coordinate_helpers.py [hunk 5]
```diff
@@ -477,6 +484,36 @@ def get_axislabel(self):
...
+    def set_auto_axislabel(self, val):
+        """
+        Render default axis labels if no explicit label is provided.
+
+        Parameters
+        ----------
+        auto_axislabel : `bool`
+            `True` if default labels will be rendered.
+        """
+        self._auto_axislabel = bool(val)
```
Comment (astrofrog) [inline astropy/visualization/wcsaxes/coordinate_helpers.py]: Looks like this method is duplicated

File: astropy/wcs/utils.py [hunk 6]
```diff
@@ -632,3 +632,39 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
     coords = cls(frame.realize_frame(data))
 
     return coords
+
+
+def local_partial_pixel_derivatives(wcs, *pixel, normalize_by_world=False):
```
Comment (astrofrog) [inline astropy/wcs/utils.py]: We should add some tests for this, even if relatively simple

File: astropy/visualization/wcsaxes/coordinate_helpers.py (hunk 4 above)
Comment (larrybradley) [reply astropy/visualization/wcsaxes/coordinate_helpers.py]: +1

**Comment by Cadair** [comment]
New reference images are in https://github.com/astropy/astropy-data/pull/76 also I wrote a script to generate them: https://github.com/astropy/astropy-procedures/pull/141

I will add a section to the narrative docs about disabling autolabel

**Comment by Cadair** [comment]
Small doc section added, let me know if you think it should be expanded.

**Comment by bsipocz** [comment]
@Cadair - please update to use the new images, I've just merged your PR in astropy-data.

**Comment by Cadair** [comment]
https://github.com/astropy/astropy-data/pull/77

**Comment by Cadair** [comment]
*slumps back into a chair*

I think this is now ready for final review.

**Comment by Cadair** [comment]
I have updated the data PR, although i haven't modified the figure for #9411

**Comment by Cadair** [comment]
Thanks everyone!!

**Pull request merged**